File: cpp/MP4File.cpp
//
//  MP4File.cpp
//  QTParse
//
//  Created by Alex Restrepo on 11/18/16.
//

#include "MP4File.h"

#include <arpa/inet.h>
#include <cassert>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <unordered_set>
#include <libgen.h>

MP4File::~MP4File() {
    if (_sourceFile) {
        fclose(_sourceFile);
    }
    _sourceFile = nullptr;
}

MP4File::MP4File(const char *path)
: _sourceFile(nullptr)
, _atoms()
{
    _sourceFile = fopen(path, "rb");
    _fileName = std::string(basename((char *)path));
    parse();
}

bool _isContainerAtom(uint32_t atomType) {
    static std::unordered_set<uint32_t> containerAtoms = {
        'moov',
        'clip',
        'udta',
        'trak',
        'clip',
        'matt',
        'edts',
        'mdia',
        'minf',
        'dinf',
        'stbl',
        'RYLO',
        'MAGE'
    };
    return containerAtoms.count(atomType) > 0;
}

off_t _fileSize(FILE *file) {
    off_t curr = ftello(file);
    
    // seek to end to get file size
    fseeko(file, 0, SEEK_END);
    off_t size = ftello(file);
    
    // seek back to original position
    fseeko(file, curr, SEEK_SET);
    
    return size;
}

#if !defined(MIN)
    #define MIN(A,B)	((A) < (B) ? (A) : (B))
#endif

size_t _copyFileData(FILE *source, FILE *destination, size_t dataSize) {
    char buffer[BUFSIZ];
    size_t copied = 0;
    size_t read = 0;
    
    while ((read = fread(buffer, 1, MIN(BUFSIZ, dataSize - copied), source))) {
        copied += fwrite(buffer, 1, read, destination);
    }
    
    return copied;
}

MP4Atom MP4File::loadAtomFromFile(FILE *file, off_t start, off_t end, MP4Atom *parent) {
    // atoms may not parse any of their data, so skip to the atom boundary.
    fseeko(file, start, SEEK_SET);
    
    // header
    // 1. size
    uint32_t size = 0;
    size_t read = fread(&size, sizeof(size), 1, file);
    size = ntohl(size);
    
    // 2. type
    uint32_t type = 0;
    read = fread(&type, sizeof(type), 1, file);
    type = ntohl(type);
    
    // extended size?
    bool hasExtendedSize = size == 1;
    uint64_t totalSize = size;
    if (hasExtendedSize) {
        read = fread(&totalSize, sizeof(totalSize), 1, file);
        totalSize = ntohll(totalSize);
    }
    
    // size must be at least 8 (4 size + 4 type)
    if(totalSize < 8
       || start + totalSize > end) {
        return MP4Atom();
    }
    
    MP4Atom atom = MP4Atom::atomWithComponents(start, hasExtendedSize, totalSize, type, parent);
    if (_isContainerAtom(type)) {
        loadContainerAtomFromFile(file, start + atom.headerSize(), atom.endOffsetInFile(), &atom, atom.children());
        
    } else {
        atom.loadPayload(file);
    }
    
    return atom;
}

void MP4File::loadContainerAtomFromFile(FILE *file, off_t start, off_t end, MP4Atom *parent, std::vector<MP4Atom> &storage) {
    off_t startOffset = start;
    
    while (startOffset < end) {
        MP4Atom atom = loadAtomFromFile(_sourceFile, startOffset, end, parent);
        if (atom.size() == 0) {
            break;
        }
        storage.push_back(atom);
        startOffset = atom.endOffsetInFile();
    }
}

void MP4File::parse() {
    if (!_sourceFile) {
        return;
    }
    loadContainerAtomFromFile(_sourceFile, 0, _fileSize(_sourceFile), nullptr, _atoms);
}

MP4Atom* MP4File::videoTrackAtom() {
    // the video track atom is located in moov > trak.
    
    for (int i = 0; i < _atoms.size(); i++) {
        if (_atoms[i].type() == 'moov') {
            std::vector<MP4Atom*> tracks = _atoms[i].childrenWithType('trak');
            for (int j = 0; j < tracks.size(); j++) {
                // there can be multiple tracks in the file, we need to find the video one.
                // for that we need to find the 'mdia' > 'hdlr' atom and extract its subtype.
                
                std::vector<MP4Atom*> mdia = tracks[j]->childrenWithType('mdia');
                if (!mdia.size()) {
                    continue;
                }
                
                std::vector<MP4Atom*> hdlr = mdia[0]->childrenWithType('hdlr');
                if (!hdlr.size()) {
                    continue;
                }
                
                // extract subtype...
                //https://developer.apple.com/library/mac/documentation/QuickTime/QTFF/QTFFChap2/qtff2.html#//apple_ref/doc/uid/TP40000939-CH204-33004
                fseek(_sourceFile, hdlr[0]->offsetInFile() + hdlr[0]->headerSize() + 8, SEEK_SET);
                uint32_t subtype = 0;
                fread(&subtype, sizeof(subtype), 1, _sourceFile);
                subtype = ntohl(subtype);
                
                if (subtype == 'vide') {
                    return tracks[j];
                }
            }
        }
    }
    
    return nullptr;
}

void MP4File::printStructure() {
    printf("\n[%s]: atoms: %lu, size %llu\n", _fileName.c_str(), _atoms.size(), calculatedSize());
    
    for (int i = 0; i < _atoms.size(); i++) {
        MP4Atom &atom = _atoms[i];
        
        if (i < _atoms.size() - 1) {
            atom.printStructureWithPadding("  ├─");
        } else {
            atom.printStructureWithPadding("  └─");
        }
    }
    
//    for (MP4Atom &atom : _atoms) {
//        atom.printStructure();
//    }
}

size_t _appendAtomHeaderToFile(MP4Atom &atom, FILE *dst) {
    size_t written = 0;
    
    uint32_t size = atom.hasExtendedSize() ? 1 : (uint32_t)atom.calculatedSize();
    size = htonl(size);
    written = fwrite(&size, 1, sizeof(uint32_t), dst);
    
    uint32_t type = atom.type();
    type = htonl(type);
    written += fwrite(&type, 1, sizeof(uint32_t), dst);
    
    if (atom.hasExtendedSize()) {
        uint64_t extendedSize = atom.calculatedSize();
        extendedSize = htonll(extendedSize);
        written += fwrite(&extendedSize, 1, sizeof(uint64_t), dst);
    }
    
    return written;
}

size_t _appendAtomToFile(MP4Atom &atom, FILE *src, FILE *dst) {
    size_t written = 0;
    
    if (atom.children().size()) {
        written += _appendAtomHeaderToFile(atom, dst);
        
        for (MP4Atom &child : atom.children()) {
            written += _appendAtomToFile(child, src, dst);
        }
        
    } else {
        if (atom.offsetInFile() < 0) {
            written += _appendAtomHeaderToFile(atom, dst);
            written += atom.appendPayloadToFile(dst);
            
        } else {
            // write the entire atom as a whole.
            fseek(src, atom.offsetInFile(), SEEK_SET);
            written += _copyFileData(src, dst, atom.calculatedSize());
        }
    }
    
    return written;
}

uint64_t MP4File::calculatedSize() {
    uint64_t size = 0;
    for (MP4Atom &atom : _atoms) {
        size += atom.calculatedSize();
    }
    
    return size;
}

bool MP4File::saveToPath(const char *path) {
    FILE *dst = fopen(path, "ab"); // append-bin
    if (!dst) {
        return false;
    }
    
    size_t written = 0;
    
    for (MP4Atom &atom : _atoms) {
        written += _appendAtomToFile(atom, _sourceFile, dst);
    }

    fclose(dst);
    return written == calculatedSize();
}

File: objc/MP4File.m
//
//  MP4File.m
//  MP4Parser
//
//  Created by Alex Restrepo on 2/11/16.
//

#import "MP4File.h"
#import "NSString+AtomType.h"

@interface MP4File ()

@property (nonatomic, copy) NSString *filePath;
@property (nonatomic, strong) NSFileHandle *readHandle;

@end

@implementation MP4File

+ (MP4File *)mp4WithFileAtPath:(NSString *)filePath {
    return [[MP4File alloc] initWithFileAtPath:filePath];
}

- (void)dealloc {
    [_readHandle closeFile];
}

- (instancetype)initWithFileAtPath:(NSString *)filePath {
    self = [super init];
    if (self) {
        self.type = 'file'; // funs
        self.headerSize = 0;
        self.offsetInFile = 0;
        if(![self parseFileAtPath:filePath]) {
            return nil;
        }
        self.size = [[self.children  lastObject] endOffsetInFile] - self.offsetInFile;
    }
    return self;
}

- (NSString *)description {
    NSString *description = [NSString stringWithFormat: @"\n[MP4File: %lu top level atoms]\n%@",
                             (unsigned long)[self.children count], [super description]];
    return description;
}

#pragma mark - Parsing
- (BOOL)parseFileAtPath:(NSString *)path {
    
    _readHandle = [NSFileHandle fileHandleForReadingAtPath:path];
    if (!_readHandle) {
        NSLog(@"Error opening file %@", path);
        return NO;
    }
    
    self.filePath = path;
    NSUInteger fileLength = [_readHandle seekToEndOfFile];
    if (fileLength) {
        self.children = [self loadContainerAtomAtOffset:0
                                              endOffset:fileLength
                                                 parent:self];
    }
    
    return [self.children count] > 0;
}

- (NSMutableArray *)loadContainerAtomAtOffset:(off_t)start
                                    endOffset:(off_t)end
                                       parent:(MP4Atom *)parent {
    off_t startPos = start;
    NSMutableArray *atoms = [[NSMutableArray alloc] init];
    while (startPos < end) {
        MP4Atom *atom = [self loadAtomAtOffset:startPos endOffset:end parent:parent];
        if (!atom) {
            //error
            return nil;
        }
        [atoms addObject:atom];
        startPos = atom.offsetInFile + atom.size;
    }
    return atoms;
}

- (MP4Atom *)loadAtomAtOffset:(off_t)start
                    endOffset:(off_t)end
                       parent:(MP4Atom *)parent {

    // atom size.
    uint32_t size = 0;
    [_readHandle he_getBytes:&size range:NSMakeRange(start, sizeof(size))];
    size = CFSwapInt32BigToHost(size);
    
    // atom type.
    uint32_t type = 0;
    [_readHandle he_getBytes:&type range:NSMakeRange(start + 4, sizeof(type))];
    type = CFSwapInt32BigToHost(type);
    
    // extended size?
    uint8_t headerSize = 8;
    uint64_t extendedSize = 0;
    if (size == 1) {
        [_readHandle he_getBytes:&extendedSize range:NSMakeRange(start + 8, sizeof(extendedSize))];
        extendedSize = CFSwapInt64BigToHost(extendedSize);
        headerSize = 16; // 4 size + 4 type + 8 extended size
        
    } else if (size < 8) {
        // this can't happen...
        NSAssert(NO, @"Invalid size atom size?");
        return nil;
    }
    
    uint64_t totalSize = size == 1 ? extendedSize : size;
    if (start + totalSize > end) {
        NSAssert(NO, @"Invalid file / Atom exceeds bounds?");
        return nil;
    }
    
    MP4Atom *atom = [MP4Atom atomForType:type];
    atom.offsetInFile = start;
    atom.size = totalSize;
    atom.headerSize = headerSize;
    atom.parent = parent;
    atom.hasExtendedSize = size == 1;
    
    // containers we're interested in
    if (
        type == 'mdia'
        || type == 'moov'
        || type == 'trak'
        || type == 'dinf'
        || type == 'minf'
        || type == 'stbl'
        || type == 'udta'
        || type == 'RYLO'
        ) {
        
        atom.children = [self loadContainerAtomAtOffset:start + atom.headerSize
                                              endOffset:[atom endOffsetInFile]
                                                 parent:atom];
        if (!atom.children) {
            return nil;
        }
        
    } else {
        [atom processData:_readHandle];
    }
    return atom;
}

- (NSArray <TRAKAtom *> *)allTracks {
    // if any, it would be inside moov
    // the type will be in moov.trak.mdia.hdlr

    // find the moov atom...
    NSArray<MP4Atom *> *moov = [self atomsWithType:'moov'];
    if ([moov count]) {
        // now, the track atoms..
        NSArray<TRAKAtom *> *tracks = [[moov firstObject] atomsWithType:'trak'];
        return tracks;
    }

    return nil;
}

- (TRAKAtom *)videoTrackAtom {
    // if any, it would be inside moov
    // the type will be in moov.trak.mdia.hdlr
    
    TRAKAtom *atom = nil;
    NSArray<TRAKAtom *> *tracks = [self allTracks];
    // when parsing the hdlr atom, it should have set a shortcut to its track...
    for (TRAKAtom *track in tracks) {
        if (track.hdlr.componentSubtype == 'vide') {
            atom = track;
            break;
        }
    }

    return atom;
}

- (int32_t)mdataOffset {
    MP4Atom *data = [self atomsWithType:'mdat'].firstObject;
    return (int32_t)([data calculatedOffsetInFile] - data.offsetInFile);
}

#pragma mark - Saving

- (BOOL)save {
    int32_t offset = [self mdataOffset];
    if (offset != 0) {
        NSArray<TRAKAtom *> *tracks = [self allTracks];
        for (TRAKAtom *track in tracks) {
            for (COAtom *coatom in [track allCoAtoms]) {
                [coatom updateEntriesWithDelta:offset];
            }
        }
    }

    // write a new tmp file, then swap/replace the old
    NSString *filePath = [[[NSUUID UUID] UUIDString] stringByAppendingPathExtension:@"mp4"];
    filePath = [NSTemporaryDirectory() stringByAppendingPathComponent:filePath];
    
    NSInteger bytesWritten = 0;
    BOOL success = NO;
    NSOutputStream *outputStream = [NSOutputStream outputStreamToFileAtPath:filePath append:NO];
    [outputStream open];
    if  ([outputStream hasSpaceAvailable]) {
        for (MP4Atom *atom in self.children) {
            bytesWritten += [self appendAtom:atom toStream:outputStream];
        }
    }
    [outputStream close];
    if (bytesWritten == [self calculatedSize]) {
        // success... replace old file
        [_readHandle closeFile];

        NSError *error = nil;
        success = [[NSFileManager defaultManager] replaceItemAtURL:[NSURL fileURLWithPath:_filePath]
                                                     withItemAtURL:[NSURL fileURLWithPath:filePath]
                                                    backupItemName:nil
                                                           options:NSFileManagerItemReplacementUsingNewMetadataOnly
                                                  resultingItemURL:nil
                                                             error:&error];
        if (success) {
            _readHandle = [NSFileHandle fileHandleForReadingAtPath:_filePath];
            [self reconcileOffsets];

        } else {
            NSLog(@"Error moving file %@", error);
        }
    }
    return success;
}

- (NSInteger)appendAtomHeader:(MP4Atom *)atom toStream:(NSOutputStream *)stream {
    NSInteger bytesWritten = 0;
    uint32_t size = atom.hasExtendedSize ? 1 : (uint32_t)[atom calculatedSize];
    size = CFSwapInt32HostToBig(size);
    bytesWritten += [stream write:(const uint8_t *)&size maxLength:sizeof(size)];
    
    uint32_t type = atom.type;
    type = CFSwapInt32HostToBig(type);
    bytesWritten += [stream write:(const uint8_t *)&type maxLength:sizeof(type)];
    
    if (atom.hasExtendedSize) {
        uint64_t extendedSize = [atom calculatedSize];
        extendedSize = CFSwapInt64BigToHost(extendedSize);
        bytesWritten += [stream write:(const uint8_t *)&extendedSize maxLength:sizeof(extendedSize)];
    }
    
    return bytesWritten;
}

- (NSInteger)appendAtom:(MP4Atom *)atom toStream:(NSOutputStream *)stream {

    NSInteger bytesWritten = 0;
    
    if ([atom.children count]) {
        bytesWritten += [self appendAtomHeader:atom toStream:stream];
        
        // then children
        for (MP4Atom *child in atom.children) {
            bytesWritten += [self appendAtom:child toStream:stream];
        }
        
    } else {
        
        // check if the atom has a valid offset, if not, it's a new atom...
        if (atom.offsetInFile < 0) {
            bytesWritten += [self appendAtomHeader:atom toStream:stream];
            bytesWritten += [atom appendContentsToStream:stream];
            
        } else {
            // there is an offset, simply copy that and append it to the stream...
            // todo, do this in pieces
//            NSData *chunk = [_data subdataWithRange:NSMakeRange(atom.offsetInFile, atom.size)];
//            bytesWritten += [stream write:chunk.bytes maxLength:[chunk length]];
            bytesWritten += [self appendDataInRange:NSMakeRange(atom.offsetInFile, atom.size)
                                           toStream:stream];
        }
    }
    
    return bytesWritten;
}

- (NSInteger)appendDataInRange:(NSRange)range toStream:(NSOutputStream *)stream {
    if (range.length == 0) {
        return 0;
    }

    const NSUInteger maxLength = 1024 * 1024 * 4; // 4 megs at a time max
    NSInteger bytesWritten = 0;
    while (bytesWritten < range.length) {
        @autoreleasepool {
            NSData *chunk = [_readHandle he_dataWithRange:NSMakeRange(range.location + bytesWritten, MIN(maxLength, range.length - bytesWritten))];
            NSInteger written = [stream write:chunk.bytes maxLength:[chunk length]];
            if (written > 0) {
                bytesWritten += written;

            } else {
                break;
            }
        }
    }

    return bytesWritten;
}

@end
